include/deal.II/lac/affine_constraints.templates.h

```diff
@@ -712,101 +712,148 @@ AffineConstraints<number>::close()
 #ifdef DEBUG
   size_type iteration = 0;
 #endif
-  bool chained_constraint_replaced = false;
+  bool              chained_constraint_replaced = false;
+  std::vector<bool> line_finalized(lines.size(), false);
   do
     {
       chained_constraint_replaced = false;
-      for (ConstraintLine &line : lines)
-        {
+      for (unsigned int line_index = 0; line_index < lines.size(); ++line_index)
+        if (line_finalized[line_index] == false)
+          {
+            ConstraintLine &line = lines[line_index];
+
 #ifdef DEBUG
-          // we need to keep track of how many replacements we do in this line,
-          // because we can end up in a cycle A->B->C->A without the number of
-          // entries growing.
-          size_type n_replacements = 0;
+            // we need to keep track of how many replacements we do in this
+            // line, because we can end up in a cycle A->B->C->A without the
+            // number of entries growing.
+            size_type n_replacements = 0;
 #endif
 
-          // loop over all entries of this line (including ones that we
-          // have appended in this go around) and see whether they are
-          // further constrained. ignore elements that we don't store on
-          // the current processor
-          unsigned int entry = 0;
-          while (entry < line.entries.size())
-            if (((local_lines.size() == 0) ||
-                 (local_lines.is_element(line.entries[entry].first))) &&
-                is_constrained(line.entries[entry].first))
+            // loop over all entries of this line (including ones that we
+            // have appended in this go around) and see whether they are
+            // further constrained. ignore elements that we don't store on
+            // the current processor.
+            //
+            // If we find that one of the entries in the current line is indeed
+            // constrained, then store the address of the line it corresponds to
+            // for further processing; this avoids having to look it up again,
+            // which requires an IndexSet lookup. If not, store a nullptr.
+            std::vector<const ConstraintLine *> sub_constraints(
+              line.entries.size(), nullptr);
```

Comment:
Would it make sense to avoid the reallocation by declaring the `std::vector` outside the do-while-loop and just resize and fill it here?